File: packages/server/src/protocol.ts
import type {
  ProviderId,
  Task,
  TaskCapabilityId,
  TaskMode,
  TaskSource,
  TaskStep
} from '@skynul/shared'

// ── Sidecar → Rust (stdout) ──────────────────────────────────────────────────

export type SidecarOutput =
  | { type: 'ready' }
  | { type: 'error'; message: string }
  | {
      type: 'task_update'
      task_id: string
      status: Task['status']
      message?: string
      step?: TaskStep
      summary?: string
    }
  | { type: 'channel_incoming'; source: TaskSource; chat_id: string; text: string }

// ── Rust → Sidecar (stdin) ───────────────────────────────────────────────────

export type SidecarInput =
  | { type: 'init'; config: InitConfig }
  | { type: 'execute'; task: TaskInput; provider: ProviderConfig }
  | { type: 'message'; task_id: string; message: string }
  | { type: 'cancel'; task_id: string }

export type InitConfig = {
  channels?: Record<string, ChannelInit>
}

export type ChannelInit = {
  enabled: boolean
  credentials: Record<string, string>
}

export type TaskInput = {
  id: string
  prompt: string
  attachments?: string[]
  mode: TaskMode
  capabilities: TaskCapabilityId[]
  maxSteps: number
  timeoutMs: number
  source?: TaskSource
}

export type ProviderConfig = {
  id: ProviderId
  model: string
  apiKey?: string
}

// ── I/O helpers ──────────────────────────────────────────────────────────────

export function writeOutput(output: SidecarOutput): void {
  process.stdout.write(`${JSON.stringify(output)}\n`)
}

export function readInput(): Promise<SidecarInput> {
  return new Promise((resolve, reject) => {
    const handler = (data: Buffer) => {
      process.stdin.removeListener('data', handler)
      try {
        resolve(JSON.parse(data.toString()))
      } catch (e) {
        reject(e)
      }
    }
    process.stdin.once('data', handler)
  })
}

export function startInputLoop(onInput: (input: SidecarInput) => void): void {
  let buffer = ''
  process.stdin.on('data', (chunk: Buffer) => {
    buffer += chunk.toString()
    const lines = buffer.split('\n')
    buffer = lines.pop() ?? ''
    for (const line of lines) {
      if (line.trim()) {
        try {
          onInput(JSON.parse(line))
        } catch {
          writeOutput({ type: 'error', message: `parse error: ${line.slice(0, 200)}` })
        }
      }
    }
  })
}

File: packages/core/src/protocol.rs
use std::collections::HashMap;
use std::process::Stdio;
use std::sync::Arc;

use serde::{Deserialize, Serialize};
use tokio::io::AsyncWriteExt;
use tokio::process::{Child, ChildStdout};
use tokio::sync::Mutex;

// ── Rust → Sidecar (stdin) ────────────────────────────────────────────────────

#[derive(Debug, Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum SidecarInput {
    Init { config: InitConfig },
    Execute { task: TaskInput, provider: ProviderConfig },
    Message { task_id: String, message: String },
    Cancel { task_id: String },
}

#[derive(Debug, Serialize, Default)]
pub struct InitConfig {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub channels: Option<HashMap<String, ChannelInit>>,
}

#[derive(Debug, Serialize)]
pub struct ChannelInit {
    pub enabled: bool,
    pub credentials: HashMap<String, String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TaskInput {
    pub id: String,
    pub prompt: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub attachments: Option<Vec<String>>,
    pub mode: String,
    pub capabilities: Vec<String>,
    pub max_steps: i64,
    pub timeout_ms: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source: Option<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ProviderConfig {
    pub id: String,
    pub model: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub api_key: Option<String>,
}

// ── Sidecar → Rust (stdout) ───────────────────────────────────────────────────

#[derive(Debug, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum SidecarOutput {
    Ready,
    Error {
        message: String,
    },
    TaskUpdate {
        task_id: String,
        status: String,
        #[serde(default)]
        message: Option<String>,
        #[serde(default)]
        step: Option<serde_json::Value>,
        #[serde(default)]
        summary: Option<String>,
    },
    ChannelIncoming {
        source: String,
        chat_id: String,
        text: String,
    },
}

// ── Handle ────────────────────────────────────────────────────────────────────

#[derive(Clone)]
pub struct SidecarHandle {
    stdin: Arc<Mutex<tokio::process::ChildStdin>>,
}

impl SidecarHandle {
    pub async fn send(&self, input: &SidecarInput) -> Result<(), std::io::Error> {
        let mut line = serde_json::to_string(input).expect("serialization cannot fail");
        line.push('\n');
        self.stdin.lock().await.write_all(line.as_bytes()).await
    }
}

// ── Spawn ─────────────────────────────────────────────────────────────────────

pub struct SpawnedSidecar {
    pub handle: SidecarHandle,
    pub child: Child,
    pub stdout: ChildStdout,
}

pub async fn spawn_sidecar(cmd: &str, args: &[&str]) -> Result<SpawnedSidecar, std::io::Error> {
    let mut child = tokio::process::Command::new(cmd)
        .args(args)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::inherit())
        .spawn()?;

    let stdin = child.stdin.take().expect("stdin must be piped");
    let stdout = child.stdout.take().expect("stdout must be piped");

    Ok(SpawnedSidecar {
        handle: SidecarHandle {
            stdin: Arc::new(Mutex::new(stdin)),
        },
        child,
        stdout,
    })
}
